llpc/util/llpcElfWriter.cpp

```diff
@@ -741,6 +741,86 @@ Result ElfWriter<Elf>::GetSectionDataBySectionIndex(
+    ElfNote*       pNewNote)       // [out] new note section
+{
+    msgpack::Document document;
+
+    auto success = document.readFromBlob(StringRef(reinterpret_cast<const char*>(pNote->pData),
+                                              pNote->hdr.descSize),
+                                              false);
+    LLPC_ASSERT(success);
+    LLPC_UNUSED(success);
+
+    auto pipeline = document.getRoot().
+        getMap(true)[Util::Abi::PalCodeObjectMetadataKey::Pipelines].getArray(true)[0];
+
+    auto registers = pipeline.getMap(true)[Util::Abi::PipelineMetadataKey::Registers].getMap(true);
+
+    auto gfxIp = pContext->GetGfxIpVersion();
+    Gfx6::PsRegConfig gfx6PsConfig;
+    Gfx9::PsRegConfig gfx9PsConfig(gfxIp);
+    Util::Abi::PalMetadataNoteEntry* pRegEntry = nullptr;
+    uint32_t psUserDataBase = 0;
+    uint32_t psUserDataCount = 0;
+    if (gfxIp.major < 9)
+    {
+        pRegEntry = reinterpret_cast<Util::Abi::PalMetadataNoteEntry*>(&gfx6PsConfig);
+        psUserDataBase = gfx6PsConfig.GetPsUserDataStart();
+        psUserDataCount = 16;
+    }
+    else
+    {
+        pRegEntry = reinterpret_cast<Util::Abi::PalMetadataNoteEntry*>(&gfx9PsConfig);
+        psUserDataBase = Gfx9::mmSPI_SHADER_USER_DATA_PS_0;
+        psUserDataCount = 32;
+    }
+
+    auto pPipelineInfo = reinterpret_cast<const GraphicsPipelineBuildInfo*>(pContext->GetPipelineBuildInfo());
+
+    for (uint32_t i = 0; i < psUserDataCount; ++i)
+    {
+        uint32_t key = psUserDataBase + i;
+        auto keyNode = registers.getDocument()->getNode(key);
+        auto keyIt = registers.find(keyNode);
+        if (keyIt != registers.end())
+        {
+            LLPC_ASSERT(keyIt->first.getUInt() == key);
+            // Reloc Descriptor user data value is consisted by DescRelocMagic | set.
+            // Max set num < 128.
+            uint32_t regValue = keyIt->second.getUInt();
+            if ((DescRelocMagic == (regValue & 0xffffff00)) && ((regValue & 0xff) < 128))
+            {
+                uint32_t set = regValue & 0xff;
+                for (uint32_t j = 0; j < pPipelineInfo->fs.userDataNodeCount; ++j)
+                {
+                    if ((pPipelineInfo->fs.pUserDataNodes[j].type == ResourceMappingNodeType::DescriptorTableVaPtr) &&
+                        (set == pPipelineInfo->fs.pUserDataNodes[j].tablePtr.pNext[0].srdRange.set))
+                    {
+                        // If it's descriptor user data, then update its offset to it.
+                        uint32_t value = pPipelineInfo->fs.pUserDataNodes[j].offsetInDwords;
+                        keyIt->second = registers.getDocument()->getNode(value);
+                        break;
+                    }
+                }
+            }
+        }
+    }
+
+    std::string blob;
+    document.writeToBlob(blob);
+    *pNewNote = *pNote;
+    auto pData = new uint8_t[blob.size() + 4]; // 4 is for additional alignment spece
```

Comment:
Is an extra word necessary here or do you want to make the size be the next multiple of 4?

Reply:
Sorry for that, which refers mergeNote. Here we just update existing note, we don't need that.  I will remove '+ 4' .